onecodex/lib/upload.py

```diff
@@ -560,220 +193,84 @@ def upload_sequence(
         # cancel_upload on later if user hits ctrl+c
-        fields = _call_init_upload(
-            filename,
-            file_size,
+
+        # upload_sequence_fileobj will call init_multipart_upload, which accepts metadata to be integrated
+        # into a newly-created Sample model.
+        # We also pass the `sample_id` and `external_sample_id`, which are typically None, to
+        # support retries of pre-uploaded samples
+        payload = _get_init_multipart_upload_payload(
+            fobj,
+            is_paired,
             metadata,
             tags,
             project,
-            samples_resource,
-            sample_id,
-            external_sample_id,
+            sample_id=sample_id,
+            external_sample_id=external_sample_id,
         )
 
+        try:
+            fields = samples_resource.init_multipart_upload(payload)
+        except requests.exceptions.HTTPError as e:
+            raise_api_error(e.response, state="init")
+
         def cancel_atexit():
             bar.canceled = True
             bar.update(1)
-            log.info("Canceled upload for sample: {}".format(fields["sample_id"]))
+            log.info("Canceled upload for sample: {}".format(payload["sample_id"]))
 
             try:
-                samples_resource.cancel_upload({"sample_id": fields["sample_id"]})
+                samples_resource.cancel_upload({"sample_id": payload["sample_id"]})
             except requests.exceptions.HTTPError as e:
                 # onecodex #298: it's possible to have this trigger after an upload has
                 # already succeeded. try to catch that instead of blowing up
                 if e.response and e.response.get("message") == "Upload already successful":
                     log.debug(
                         "Fail to cancel sample {}, upload already successful".format(
-                            fields["sample_id"]
+                            payload["sample_id"]
                         )
                     )
                 else:
                     raise
 
         atexit_register(cancel_atexit)
 
-        # if the upload via init_upload fails, upload_sequence_fileobj will call
-        # init_multipart_upload, which accepts metadata to be integrated into a newly-created
-        # Sample model. if the s3 intermediate route is used, two Sample models will ultimately
-        # exist on mainline: the failed fastx-proxy upload and the successful s3 intermediate.
-        # We also pass the `sample_id` and `external_sample_id`, which are typically None, to
-        # support retries of pre-uploaded samples
-        retry_fields = _make_retry_fields(
-            filename,
-            metadata,
-            tags,
-            project,
-            sample_id=sample_id,
-            external_sample_id=external_sample_id,
-        )
+        if is_paired:
+            # 2 files to upload
+            _upload_sequence_fileobj(fobj.r1, fobj.r1.filename, fields, samples_resource)
```

Comment:
Note: We should make sure we clean these up on the `/cancel` call on the backend.